Add ability to skip read-only replicas for INSERT into Distributed engine

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add ability to skip read-only replicas for INSERT into Distributed engine (Controlled with `distributed_insert_skip_read_only_replicas` setting, by default OFF - backward compatible)

Sometimes replica may be readonly for a long time, and this will make
some INSERT queries fail, but it does not make sense to INSERT into
readonly replica, so let's ignore them.

But note, that this will require to extend TableStatus (not extend, but
introduce new version), that will have is_readonly field.

Also before background INSERT into Distributed does not uses
getManyChecked() which means that they do not request TableStatus
packet, while now they would, though this is minor (just a note).

azat committed:
Do not try to INSERT into readonly replicas for Distributed engine

Sometimes replica may be readonly for a long time, and this will make
some INSERT queries fail, but it does not make sense to INSERT into
readonly replica, so let's ignore them.

But note, that this will require to extend TableStatus (not extend, but
introduce new version), that will have is_readonly field.

Also before background INSERT into Distributed does not uses
getManyChecked() which means that they do not request TableStatus
packet, while now they would, though this is minor (just a note).

v2: Add a note about max_replica_delay_for_distributed_queries for INSERT
v3: Skip read-only replicas for async INSERT into Distributed
v4: Remove extra @insert parameter for ConnectionPool*::get*
    It make sense only when the table name had passed --
    ConnectionPoolWithFailover::getManyChecked()
v5: rebase on top LoggerPtr
v6: rebase
v7: rebase
v8: move TryResult::is_readonly into the end

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mark Replicated table as readonly in case of static disk (read-only/write-once)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: cover filtering out read-only replicas during INSERT into Distributed

v2: replace without s3_plain disk
v3: add check for DistributedConnectionReadOnlyReplica
v4: set insert_keeper_max_retries=0
v5: do not set insert_keeper_max_retries=0
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #59176 from azat/ClickHouse:dist/readonly-replicas -->
*This is an automated comment for commit a88c7cc65e2bfba8f72f2dc280743b8017bb4fc3 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59176/a88c7cc65e2bfba8f72f2dc280743b8017bb4fc3/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59176/a88c7cc65e2bfba8f72f2dc280743b8017bb4fc3/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

azat committed:
tests: adjust 02537_distributed_loosing_files_after_exception

Now, since there is connection probing, the error changed from
UNKNOWN_TABLE to ALL_CONNECTION_TRIES_FAILED, though I'm not sure that
this good...

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Test failures does not looks related:

>Stateless tests (debug) [2/5] — fail: 1, passed: 1164, skipped: 10

- `01459_manual_write_to_replicas`

>Stateless tests (debug) [3/5] — fail: 2, passed: 1237, skipped: 22

- `01508_partition_pruning_long`
- `02125_lz4_compression_bug`

>Upgrade check (debug) — Logical error thrown (see clickhouse-server.log or logical_errors.txt)
>Upgrade check (tsan) — Logical error thrown (see clickhouse-server.log or logical_errors.txt)

- #58319

devcrafter commented on src/Client/ConnectionEstablisher.cpp:
```diff
@@ -72,6 +74,16 @@ void ConnectionEstablisher::run(ConnectionEstablisher::TryResult & result, std::
         result.is_usable = true;
 
+        bool is_readonly = table_status_it->second.is_readonly;
+        if (insert && is_readonly)
```
Passing `insert` flag here is too intrusive. Logically, connection is almost nothing to do with operations on top of it (it only provides connections with some properties). I'd suggest adding `read_only` flag into `TryResult`, so you can check in upper level code

azat replied:
Make sense, fixed in c0e06cbc11269e7841ca3a0e863d5a1ff8cce6d6

azat committed:
Move insert flag from ConnectionEstablisher to TryResult

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>Stateless tests (tsan, s3 storage) [3/5] — fail: 1, passed: 1223, skipped: 38

- `02981_vertical_merges_memory_usage`

```
Timeout exceeded while receiving data from server. Waited for 300 seconds, timeout is 300 seconds
```

devcrafter reviewed (changes requested):
I think, the new behavior requires a setting. In general, writing into a currently read-only replication table is covered by insert retries.

The question is what semantic we want to bring with the new setting
(1) - prefer insert into currently non-read-only replicas
or 
(2) - use only non-read-only replicas

AFAIU, the current implementation is about (1)

devcrafter reviewed (commented):
I think, the new behavior requires a setting. In general, writing into a currently read-only replication table is covered by insert retries.

The question is what semantic we want to bring with the new setting
(1) - prefer insert into currently non-read-only replicas
or 
(2) - use only non-read-only replicas

AFAIU, the current implementation is about (1)

devcrafter reviewed (commented):
I think, the new behavior requires a setting. In general, writing into a currently read-only replication table is covered by insert retries.

The question is what semantic we want to bring with the new setting
(1) - prefer insert into currently non-read-only replicas
or 
(2) - use only non-read-only replicas

AFAIU, the current implementation is about (1)

azat commented:
> I think, the new behavior requires a setting. In general, writing into a currently read-only replication table is covered by insert retries.

My thinking was that since it only prefers non-read-only replicas, there is no need in a separate setting, but I'm OK with adding one.

>The question is what semantic we want to bring with the new setting
>(1) - prefer insert into currently non-read-only replicas
>or
>(2) - use only non-read-only replicas

And yes, if there will be a setting then we can add a enum for it, i.e. `prefer`, `no`, `yes` or similar

> AFAIU, the current implementation is about (1)

Yes.

devcrafter commented:
> > AFAIU, the current implementation is about (1)
> 
> Yes.

Let's keep it. Doubt that we need (2) actually

azat commented:
> Let's keep it. Doubt that we need (2) actually

Do you still think that extra setting make sense?

devcrafter commented:
> > Let's keep it. Doubt that we need (2) actually
> 
> Do you still think that extra setting make sense?

Hard to say, but at least there will be a way to control behavior. I guess there could be cases which benefit or not from this behavior (I guess it's mostly about load distribution). For example, if several long-running insert queries are thrown while some replicas will reconnect to zk, the inserts will be assigned to non-read-only replicas, i.e. load will not be distributed well.

When the user definitely knows that some replicas will be read-only for a long time, then it is worth excluding read-only replicas upfront. But then, we're talking about option (2) 🤔

azat committed:
Introduce distributed_insert_prefer_non_readonly_replica setting

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Simpler and less error prone PoolWithFailoverBase::TryResult::reset()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove insert argument for ConnectionPoolWithFailover

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Core/ProtocolDefines.h:
```diff
@@ -76,13 +76,16 @@ static constexpr auto DBMS_MIN_REVISION_WITH_SPARSE_SERIALIZATION = 54465;
 /// later is just a number for server version (one number instead of commit SHA)
 /// for simplicity (sometimes it may be more convenient in some use cases).
-static constexpr auto DBMS_TCP_PROTOCOL_VERSION = 54466;
+static constexpr auto DBMS_TCP_PROTOCOL_VERSION = 54467;
```
Sure, otherwise server will never use it.

azat committed:
Remove superior check for empty pools

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Convert distributed_insert_replicas_preferences into enum

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Revert extra changed lines

That was the changes that looks OK for the initial patchset version, but
not now, after all review comments.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Client/ConnectionPoolWithFailover.h:
```diff
@@ -64,18 +64,21 @@ class ConnectionPoolWithFailover : public IConnectionPool, private PoolWithFailo
         const Settings & settings,
         PoolMode pool_mode,
         const QualifiedTableName & table_to_check,
+        bool insert,
```
Ok, I will move this logic out from the connection pool.

azat commented on src/Client/ConnectionEstablisher.cpp:
```diff
@@ -87,6 +95,7 @@ void ConnectionEstablisher::run(ConnectionEstablisher::TryResult & result, std::
 
             LOG_TRACE(log, "Server {} has unacceptable replica delay for table {}.{}: {}", result.entry->getDescription(), table_to_check->database, table_to_check->table, delay);
             ProfileEvents::increment(ProfileEvents::DistributedConnectionStaleReplica);
+            return;
```
Neither do I, I'd reverted this hunk.

azat commented:
Okay, since you are also struggling about new setting for this functionality, I've added enum, `distributed_insert_replicas_preferences=PREFER_NON_READ_ONLY/NO_PREFERENCES`.

And also addressed all of your comments.

azat committed:
Add documentation for distributed_insert_replicas_preferences

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Test failures does not looks related:

>Stress test (debug) — Killed by signal (in clickhouse-server.log)
>Stress test (tsan) — Killed by signal (in clickhouse-server.log)

- https://github.com/ClickHouse/ClickHouse/issues/60349

devcrafter commented on src/Common/PoolWithFailoverBase.h:
```diff
@@ -73,19 +73,14 @@ class PoolWithFailoverBase : private boost::noncopyable
                                 /// if table is not present on remote peer, -> it'll be false
         bool is_up_to_date = false; /// If true, the entry is a connection to up-to-date replica
                                     /// Depends on max_replica_delay_for_distributed_queries setting
+        bool is_readonly = false;   /// Table is in read-only mode, INSERT can ignore such replicas.
```
Let's add the new field to the end of the structure, so the field will be declared in the same sequence as their serialization.

devcrafter commented on src/Storages/Distributed/DistributedSink.cpp:
```diff
@@ -372,10 +370,11 @@ DistributedSink::runWritingJob(JobReplica & job, const Block & current_block, si
-
+                    /// NOTE: INSERT will also take into account max_replica_delay_for_distributed_queries
+                    /// (anyway fallback_to_stale_replicas_for_distributed_queries=true by default)
+                    auto results = shard_info.pool->getManyChecked(timeouts, settings, PoolMode::GET_ONE, storage.remote_storage.getQualifiedName());
```
Unfortunately, it won't work. Here we'll get only one connection in `results` since `PoolMode::GET_ONE`, and it can be to read-only replica. So, it looks like having a setting `skip_read_only_replicas` is the simplest way to implement avoidance of the read-only replicas during distributed insert.
Yes, it's different from `prefer_non_read_only_replicas`, but it's a bit simpler from an implementation standpoint and has more deterministic outcome (not completely though, since replica can become read-only anytime).

So, I'd suggest adding just one new setting `skip_read_only_replicas` instead of introducing enum setting.

P.S. I apologize for confusion if any, - it seems we're going back and forth, but at least now I have a more clear POV.

azat replied:
>Unfortunately, it won't work. Here we'll get only one connection in results since PoolMode::GET_ONE, and it can be to read-only replica

That is correct, in this version of the patch it does not work, while it works in one of the intermediate versions, when the sorting of the shards was done inside the pool.

>So, I'd suggest adding just one new setting skip_read_only_replicas instead of introducing enum setting.

OK, let's go this way.

azat replied:
Done

azat committed:
Replace distributed_insert_replicas_preferences with distributed_insert_skip_read_only_replicas

v2: fix test
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

devcrafter commented on tests/queries/0_stateless/02980_dist_insert_readonly_replica.sql.j2:
```diff
@@ -0,0 +1,84 @@
+insert into dist_single_no_internal_replication settings distributed_foreground_insert=1, distributed_insert_skip_read_only_replicas=0 values (0); -- { serverError TABLE_IS_READ_ONLY }
+insert into dist_single_no_internal_replication settings distributed_foreground_insert=1, distributed_insert_skip_read_only_replicas=1 values (0); -- { serverError TABLE_IS_READ_ONLY }
+
+-- for internal_replicatio==true, it does
```
Don't really understand why `internal_replication` affects the outcome. Only the reason for failure is the read-only remote table, isn't it?

azat replied:
>Don't really understand why internal_replication affects the outcome.

Because in case of `internal_replication=false` it cannot skip any replicas, since it need to duplicate the data to all replicas on the initiator.

devcrafter replied:
Please make a comment to make it obvious

devcrafter commented on tests/queries/0_stateless/02980_dist_insert_readonly_replica.sql.j2:
```diff
@@ -0,0 +1,84 @@
+create table dist_batch as shard_0.data engine=Distributed('test_cluster_two_replicas_different_databases_internal_replication', '', 'data') settings background_insert_batch=1;
+system stop distributed sends dist_batch;
+
+create table dist_single_no_internal_replication as shard_0.data engine=Distributed('default', 'shard_1', 'data');
```
Can we name table with `read_only` in it. So, it'd clear from name if we refer to read-only table

devcrafter commented on tests/queries/0_stateless/02980_dist_insert_readonly_replica.sql.j2:
```diff
@@ -0,0 +1,84 @@
+insert into dist_single settings distributed_foreground_insert=1, distributed_insert_skip_read_only_replicas=0 values (0); -- { serverError TABLE_IS_READ_ONLY }
+insert into dist_single settings distributed_foreground_insert=1, distributed_insert_skip_read_only_replicas=1 values (0); -- { serverError ALL_CONNECTION_TRIES_FAILED }
+
+-- min_insert_block_size_rows is not enough, since replica will be selected
```
AFAIU, it can't be guaranteed that we'll try to use connection to read-only replica during connection selection. So, checking event profile `DistributedConnectionReadOnlyReplica` will fail eventually. Better check it reliably with failed queries above.

azat replied:
30 blocks usually enough to hit failed replica at least once

azat replied:
OK, with `load_balancing='round_robin'` this part should not be flaky at all

azat committed:
Simplify condition in PoolWithFailoverBase

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Rename to DistributedConnectionSkipReadOnlyReplica

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove distributed_insert_skip_read_only_replicas from settings history

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Core/SettingsChangesHistory.h:
```diff
@@ -114,6 +114,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"async_insert_busy_timeout_max_ms", 200, 200, "The minimum value of the asynchronous insert timeout in milliseconds; async_insert_busy_timeout_ms is aliased to async_insert_busy_timeout_max_ms"},
               {"async_insert_busy_timeout_increase_rate", 0.2, 0.2, "The exponential growth rate at which the adaptive asynchronous insert timeout increases"},
               {"async_insert_busy_timeout_decrease_rate", 0.2, 0.2, "The exponential growth rate at which the adaptive asynchronous insert timeout decreases"},
+              {"distributed_insert_skip_read_only_replicas", false, true, "If true, INSERT into Distributed will skip read-only replicas"},
```
AFAIR otherwise the compatibility test falis, let's try without (anyway the new value is incorrect)

azat committed:
Cleanup test

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix settings history for distributed_insert_skip_read_only_replicas

CI: https://s3.amazonaws.com/clickhouse-test-reports/59176/fa0c47a42f4187e8284e19d1cd008a0a76c38caf/fast_test.html
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add comments for 02980_dist_insert_readonly_replica

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make 02980_dist_insert_readonly_replica test more stable (by using round_robin LB)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Do not use default (removed) cluster in 02980_dist_insert_readonly_replica

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>
